### PR TITLE
ErProof class with LFSC output

### PR DESCRIPTION
**Pull request by alex-ozdemir**
* Created a TraceCheckProof class
   * parsable from text
* Created an ErProof class
   * constructible from a TraceCheckProof
   * writable as LFSC
* A bunch of unit tests

You can run all the proof unit tests with `ctest -R unit/proof`.

This diff is rather long diff -- sorry about that. It was shorted, but I decided that more testing was a good idea after finding a bug.

### CODE REVIEW

File: src/proof/er/er_proof.cpp [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+      prop::SatLiteral pivot = pivots[i];
+      os << "(" << (pivot.isNegated() ? 'Q' : 'R') << " _ _ ";
+    }
+
+    // Print resolution start
+    writeIdForClauseProof(os, chain[0]);
+    os << " ";
+
+    // Print resolution closings (forward order)
+    for (size_t i = 0, n = pivots.size(); i < n; ++i)
```
Comment (mpreiner) [inline src/proof/er/er_proof.cpp]: Maybe use `for (auto pivot : pivots) { ... }`

File: src/proof/er/er_proof.cpp [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    {
+      prop::SatVariable pivotVar = pivots[i].getSatVariable();
+      TraceCheckIdx clauseId = chain[i + 1];
+      writeIdForClauseProof(os, clauseId);
+      os << " ";
+      if (newVariables.find(pivotVar) != newVariables.end())
```
Comment (mpreiner) [inline src/proof/er/er_proof.cpp]: `ContainsKey(newVariables, pivotVar)`

File: src/proof/er/er_proof.cpp [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+}
+
+namespace {
+/**
+ * Resolves two clauses
+ *
+ * @param dest one of the inputs, and the output too
+ * @param src the other input
+ *
+ * @return the unique literal that was resolve on, polarizes per its
```
Comment (mpreiner) [inline src/proof/er/er_proof.cpp]:
```suggestion
 * @return the unique literal that was resolved on, polarizes per its
```

Comment (mpreiner) [reply src/proof/er/er_proof.cpp]: Not sure if I fully understand the polarizes part here. Maybe rephrase and make clearer.

File: src/proof/er/er_proof.cpp [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+
+  // Finally, close the parentheses!
+  std::fill_n(std::ostream_iterator<char>(os), parenCount, ')');
+}
+
+namespace {
+/**
+ * Resolves two clauses
+ *
+ * @param dest one of the inputs, and the output too
```
Comment (mpreiner) [inline src/proof/er/er_proof.cpp]: So this is input and output param? The result will be stored in `dest`?

File: src/proof/er/er_proof.cpp [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+}
+}  // namespace
+
+std::vector<prop::SatLiteral> ErProof::computePivotsForChain(
+    const std::vector<TraceCheckIdx>& chain) const
+{
+  std::vector<prop::SatLiteral> pivots;
+
+  std::unordered_set<prop::SatLiteral, prop::SatLiteralHashFunction>
+      runningClause;
```
Comment (mpreiner) [inline src/proof/er/er_proof.cpp]: `runningClause(runningClause.end(), first.begin(), first.end());`

File: src/proof/er/er_proof.h [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+struct ErDefinition
+{
+  ErDefinition(prop::SatVariable newVariable,
+               prop::SatLiteral oldLiteral,
+               std::vector<prop::SatLiteral>&& otherLiterals)
+      : d_newVariable(newVariable),
+        d_oldLiteral(oldLiteral),
+        d_otherLiterals(otherLiterals)
+  {
+    // That's all!
```
Comment (mpreiner) [inline src/proof/er/er_proof.h]: Comment not necessary ;-)

File: src/proof/er/er_proof.h [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+ * this new clause.
+ */
+struct TraceCheckLine
+{
+  TraceCheckLine(TraceCheckIdx idx,
+                 std::vector<prop::SatLiteral>&& clause,
+                 std::vector<TraceCheckIdx>&& chain)
+      : d_idx(idx), d_clause(clause), d_chain(chain)
+  {
+    // That's all
```
Comment (mpreiner) [inline src/proof/er/er_proof.h]: Ditto.

File: src/proof/er/er_proof.h [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+  }
+
+  // The index of the new clause
+  TraceCheckIdx d_idx;
+  // The new clause
+  std::vector<prop::SatLiteral> d_clause;
+  /**
+   * Indices of clauses which must be resolved to produce this new clause
```
Comment (mpreiner) [inline src/proof/er/er_proof.h]:
```suggestion
   * Indices of clauses which must be resolved to produce this new clause.
```

File: src/proof/er/er_proof.h [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+/**
+ * A TRACECHECK proof -- just a list of lines
+ */
+struct TraceCheckProof
+{
+  static TraceCheckProof fromText(std::istream& in);
+  TraceCheckProof() : d_lines()
+  {
+    // That's all
```
Comment (mpreiner) [inline src/proof/er/er_proof.h]: Ditto.

File: src/proof/er/er_proof.h [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+  const TraceCheckProof& getTraceCheckProof() const { return d_tracecheck; }
+
+ private:
+  /**
+   * Creates an empy ErProof.
+   */
+  ErProof() : d_inputClauseIds(), d_definitions(), d_tracecheck()
+  {
+    // That's all!
```
Comment (mpreiner) [inline src/proof/er/er_proof.h]: Ditto.

**Comment by alex-ozdemir** [comment]
Thanks for reviewing this! I just saw it now, and will respond to your comments soon.

File: src/proof/er/er_proof.cpp [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief ER Proof Format
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Sounds good.

File: src/proof/er/er_proof.cpp [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+  r = mkstemp(formulaFilename);
+  AlwaysAssert(r > 0);
+  close(r);
+  r = mkstemp(dratFilename);
+  AlwaysAssert(r > 0);
+  close(r);
+  r = mkstemp(tracecheckFilename);
+  AlwaysAssert(r > 0);
+  close(r);
+
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: They don't :frowning: . Good catch!

File: src/proof/er/er_proof.cpp [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+  // Step zero, save input clause Ids for future printing
+  std::transform(usedClauses.begin(),
+                 usedClauses.end(),
+                 std::inserter(d_inputClauseIds, d_inputClauseIds.end()),
+                 [](const std::pair<ClauseId, prop::SatClause>& pair) {
+                   return pair.first;
+                 });
+
+  // Step one, verify the formula starts the proof
+  for (size_t i = 0, n = usedClauses.size(); i < n; ++i)
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Sounds good! I'll use `CVC4_ASSERTIONS`.

File: src/proof/er/er_proof.cpp [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+    {
+      Assert(usedClauses[i].second[j] == d_tracecheck.d_lines[i].d_clause[j]);
+    }
+  }
+
+  // Step two, identify definitions. They correspond to lines that follow the
+  // input lines, are in bounds, and have no justifying chain.
+  for (size_t i = usedClauses.size();
+       i < d_tracecheck.d_lines.size()
+       && d_tracecheck.d_lines[i].d_chain.size() == 0;
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Sure.

File: src/proof/er/er_proof.cpp [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+      )
+  {
+    Assert(d_tracecheck.d_lines[i].d_clause.size() > 0);
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Good idea!

File: src/proof/er/er_proof.cpp [hunk 16]
```diff
@@ -0,0 +1,368 @@
...
+    Assert(!d_tracecheck.d_lines[i].d_clause[0].isNegated());
+
+    // Get the new variable of the definition -- the first variable of the
+    // first clause
+    prop::SatVariable newVar =
+        d_tracecheck.d_lines[i].d_clause[0].getSatVariable();
+
+    // The rest of the literals in the clause of the 'other literals' of the def
+    std::vector<prop::SatLiteral> otherLiterals;
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Oh neat. I didn't know about that constructor. Thanks!

File: src/proof/er/er_proof.cpp (hunk 1 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Alas, we need `i` within the loop in order to index `chain` properly. If C++ had a zip, that would work here.

File: src/proof/er/er_proof.cpp (hunk 2 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Sounds good!

File: src/proof/er/er_proof.cpp (hunk 4 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Yes. I'll add a note to make this more clear.

File: src/proof/er/er_proof.cpp (hunk 3 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Sounds good!

File: src/proof/er/er_proof.cpp (hunk 5 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Great!

File: src/proof/er/er_proof.h (hunk 6 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: :+1:

File: src/proof/er/er_proof.h (hunk 7 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: :+1:

File: src/proof/er/er_proof.h (hunk 8 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: :+1:

File: src/proof/er/er_proof.h (hunk 9 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: :+1:

File: src/proof/er/er_proof.h [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+{
+ public:
+  /**
+   * Construct an ER proof from a DRAT proof, using drat2er
+   *
+   * @param usedClauses The CNF formula that we're deriving bottom from.
+   * @param dratBinary  The DRAT proof from the SAT solver, as a binary stream.
+   */
+  static ErProof fromBinaryDratProof(
+      const std::vector<std::pair<ClauseId, prop::SatClause>>& usedClauses,
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: Sounds good!

File: src/proof/er/er_proof.h (hunk 10 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: :+1:

File: src/proof/er/er_proof.h [hunk 18]
```diff
@@ -0,0 +1,216 @@
+/*********************                                                        */
+/*! \file er_proof.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: :+1:

File: src/proof/er/er_proof.cpp [hunk 19]
```diff
@@ -0,0 +1,368 @@
+/*********************                                                        */
+/*! \file er_proof.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: :+1:

File: src/proof/er/er_proof.h [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+  std::vector<prop::SatLiteral> d_otherLiterals;
+};
+
+// For representing a clause's index within a TRACECHECK proof.
+using TraceCheckIdx = size_t;
+
+/**
+ * A single line in a TRACECHECK proof.
+ *
+ * Consists of a new clause, the index for that clause, and the indices for
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: I think that the comma is more accurate here -- the noun phrase is meant to be a list of three items.

File: src/proof/er/er_proof.h [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+ * preceeding clauses which can be combined in a resolution chain to produce
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: :+1:

File: src/proof/er/er_proof.h [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+   * @param dratBinary  The DRAT proof from the SAT solver, as a binary stream.
+   */
+  static ErProof fromBinaryDratProof(
+      const std::vector<std::pair<ClauseId, prop::SatClause>>& usedClauses,
+      const std::string& dratBinary);
+
+  /**
+   * Construct an ER proof from a TRACECHECK ER proof
+   *
+   * This basically just identifies groups lines which correspond to
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: :+1:

File: src/proof/er/er_proof.h [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+  const std::vector<ErDefinition>& getDefinitions() const
+  {
+    return d_definitions;
+  }
+
+  const TraceCheckProof& getTraceCheckProof() const { return d_tracecheck; }
+
+ private:
+  /**
+   * Creates an empy ErProof.
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: :+1:

File: src/proof/er/er_proof.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/proof/er/er_proof.h]: Oh, it looks like I got a bit confused by the order here :) Maybe it would be slightly better to say "The index of a new clause, the literals of that clause, and the indices..." (to match the order of the arguments and the order in the TRACECHECK format).

File: test/unit/proof/er_proof_black.h [hunk 24]
```diff
@@ -0,0 +1,421 @@
...
+ * @return a string without whitespace
+ */
+std::string filterWhitespace(const std::string& s)
+{
+  std::string out;
+  std::copy_if(s.cbegin(), s.cend(), std::inserter(out, out.end()), [](char c) {
+    return !std::isspace(c);
+  });
+  return out;
+}
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply test/unit/proof/er_proof_black.h]: Sure! I'll give it a shot.

File: src/proof/er/er_proof.cpp [hunk 25]
```diff
@@ -0,0 +1,368 @@
...
+
+ErProof::ErProof(
+    const std::vector<std::pair<ClauseId, prop::SatClause>>& usedClauses,
+    TraceCheckProof&& tracecheck)
+    : d_inputClauseIds(), d_definitions(), d_tracecheck(tracecheck)
+{
+  // Step zero, save input clause Ids for future printing
+  std::transform(usedClauses.begin(),
+                 usedClauses.end(),
+                 std::inserter(d_inputClauseIds, d_inputClauseIds.end()),
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Cool. I didn't know that existed!

File: src/proof/er/er_proof.h [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+      const std::string& dratBinary);
+
+  /**
+   * Construct an ER proof from a TRACECHECK ER proof
+   *
+   * This basically just identifies groups lines which correspond to
+   * definitions, and extracts them.
+   *
+   * @param usedClauses The CNF formula that we're deriving bottom from.
+   * @param tracecheck  The TRACECHECK proof, as a stream.
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: Hmm. Isn't that what the type is for?

**Comment by alex-ozdemir** [comment]
I think I've address both of your sets of comments so far. Let me know if this is not the case!

File: src/proof/er/er_proof.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.h]: Good idea!

File: src/proof/er/er_proof.cpp (hunk 1 above)
Comment (mpreiner) [reply src/proof/er/er_proof.cpp]: Sure! Missed that `i` in the loop.

File: src/proof/er/er_proof.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: Correction: I actually used `Configuration::isAssertionBuild()`

File: test/unit/proof/whitespace.h [hunk 27]
```diff
@@ -0,0 +1,34 @@
+/*********************                                                        */
+/*! \file whitespace.h
```
Comment (4tXJ7f) [inline test/unit/proof/whitespace.h]: Can we maybe change the filename to `utils.h` s.t. we can (potentially) add other helper functions here?

File: src/proof/er/er_proof.cpp (hunk 25 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/proof/er/er_proof.cpp]: One of the best and the worst features of C++ is that you never stop learning about the language :D

File: src/proof/er/er_proof.h (hunk 26 above)
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/proof/er/er_proof.h]: Yeah, you are right. I think it is fine as-is.

File: src/proof/er/er_proof.cpp [hunk 28]
```diff
@@ -0,0 +1,378 @@
...
+      "ER proof production requires drat2er.\n"
+      "Run contrib/get-drat2er, reconfigure with --drat2er, and rebuild");
+#endif
+
+  // Parse the resulting TRACECHECK proof into an ER proof.
+  std::ifstream tracecheckStream(tracecheckFilename);
+  ErProof proof(usedClauses, TraceCheckProof::fromText(tracecheckStream));
+  tracecheckStream.close();
+
+  unlink(tracecheckFilename);
```
Comment (4tXJ7f) [inline src/proof/er/er_proof.cpp]: I would suggest to unlink the files as soon as they are opened. Like that, they are deleted as soon as CVC4 terminates, no matter the reason (see #2834).

File: src/proof/er/er_proof.cpp [hunk 29]
```diff
@@ -0,0 +1,378 @@
...
+    prop::SatVariable newVar = c[0].getSatVariable();
+
+    // The rest of the literals in the clause of the 'other literals' of the def
+    std::vector<prop::SatLiteral> otherLiterals{++c.begin(), c.end()};
+
+    size_t nLinesForThisDef = 2 + otherLiterals.size();
+    // Look at the negation of the second literal in the second clause to get
+    // the old literal
+    d_definitions.emplace_back(newVar,
+                               ~d_tracecheck.d_lines[i + 1].d_clause[1],
```
Comment (4tXJ7f) [inline src/proof/er/er_proof.cpp]: Could it ever happen that `i + 1` is out of bounds here? E.g. if a proof only has definitions for some weird reason?

File: src/proof/er/er_proof.cpp (hunk 28 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: I see what you mean. Good idea, will do.

Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: To ensure that the file isn't removed while the program is running, I delayed our file closes.

File: test/unit/proof/whitespace.h (hunk 27 above)
Comment (alex-ozdemir) [reply test/unit/proof/whitespace.h]: Sounds good.

File: src/proof/er/er_proof.cpp (hunk 29 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: It turns out that definition actually manifest as `2 + n` lines in the TRACECHECK proof, where `n` is non-negative. Thus this line is still working with the same definition. That being said, if the definition was ill-formed (too short), this would be a segfault, and I should assert that the index is in bounds.

**Comment by alex-ozdemir** [comment]
I think I've responded to your second review, Andres. Let me know if I missed anything.

File: src/proof/er/er_proof.cpp (hunk 28 above)
Comment (4tXJ7f) [reply src/proof/er/er_proof.cpp]: The file itself is only removed after the last program that has it open is terminated [[0]](http://pubs.opengroup.org/onlinepubs/9699919799/functions/unlink.html). However, I just realized there is an issue with it because while it is guaranteed that the file is not deleted, the file is not accessible through the unlinked path anymore. Let's leave it as it is for now.

File: src/proof/er/er_proof.cpp (hunk 29 above)
Comment (4tXJ7f) [reply src/proof/er/er_proof.cpp]: Yes, it would be great to assert it because it is actually not guaranteed to segfault but could silently read garbage (hopefully this would be caught later though).

**Comment by 4tXJ7f** [comment]
Did you forget to push your changes by any chance? :) From your answers, it sounded like you are going to do the renaming and add the assertion.

**Comment by alex-ozdemir** [comment]
Whoops! My mistake -- I did forget to push to this branch. Done now.

File: src/proof/er/er_proof.cpp (hunk 29 above)
Comment (alex-ozdemir) [reply src/proof/er/er_proof.cpp]: :+1:
